### PR TITLE
Serialize exercise documents

### PR DESCRIPTION
**Pull request by ErikSchierboom**

### CODE REVIEW

File: app/commands/parse_markdown.rb [hunk 1]
```diff
@@ -32,18 +32,18 @@ def sanitized_html
 
   memoize
   def raw_html
-    renderer = Renderer.new(options: [:UNSAFE], nofollow_links: nofollow_links)
-    html = CommonMarker.render_doc(
+    doc = CommonMarker.render_doc(
```
Comment (ErikSchierboom) [inline app/commands/parse_markdown.rb]: This is just a naming refactoring. It was previously named `html`, but it is actually containing the parsed document.

File: app/commands/parse_markdown.rb [hunk 2]
```diff
@@ -32,18 +32,18 @@ def sanitized_html
...
     )
-    renderer.render(html)
+    renderer = Renderer.new(options: [:UNSAFE], nofollow_links: nofollow_links)
+    renderer.render(doc)
   end
 
   memoize
   def preprocessed_text
-    text.gsub(/^`{3,}(.*?)`{3,}\s*$/m) { "\n#{$&}\n" }
+    text.gsub(/^`{3,}(.*?)`{3,}\s*$/m) { "\n#{Regexp.last_match(0)}\n" }
```
Comment (ErikSchierboom) [inline app/commands/parse_markdown.rb]: This is not me, this must be Rubocop.

File: app/serializers/serialize_exercise_instructions.rb [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+      join
+  end
+
+  def general_hints
+    hints["general"].to_a
+  end
+
+  memoize
+  def hints
+    hints_doc.each_cons(2).each_with_object({}) do |(header, list), hints|
```
Comment (iHiD) [inline app/serializers/serialize_exercise_instructions.rb:47]:
```suggestion
    # Hints documents are structured as a series of headers and lists:
    # e.g.
    #  ## Header
    #  - foo
    #  - bar
    hints_doc.each_cons(2).each_with_object({}) do |(header, list), hints|
```

File: app/serializers/serialize_exercise_instructions.rb [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      next unless header.type == :header
```
Comment (iHiD) [inline app/serializers/serialize_exercise_instructions.rb:50]:
```suggestion
      # TODO: Add an issue to the relevant track via a async job 
      # if any of these are invalid
      next unless header.type == :header
```

File: app/serializers/serialize_exercise_instructions.rb [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    end
+  end
+
+  def tasks
+    instructions_doc.each.
+      drop_while { |node| node.type != :header }.
+      chunk_while { |_, nxt| nxt.type != :header }.
+      each_with_object([]) do |nodes, tasks|
+        task_title = parse_title(nodes.first)
+        task_text = nodes.drop(1).each.map(&:to_html).join.strip
```
Comment (ErikSchierboom) [inline app/serializers/serialize_exercise_instructions.rb]: This has a small hidden bug, as just concatenating the HTML results in spacing being lost. I have to check how the HTML would be rendered when the elements are combined.

File: app/serializers/serialize_exercise_instructions.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+      heading = parse_title(header).downcase
+      hints[heading] = list.each.map { |list_item| list_item.each.first.to_html }
+    end
+  end
+
+  def tasks
+    instructions_doc.each.
+      drop_while { |node| node.type != :header }.
+      chunk_while { |_, nxt| nxt.type != :header }.
+      each_with_object([]) do |nodes, tasks|
```
Comment (iHiD) [inline app/serializers/serialize_exercise_instructions.rb]:
```suggestion
      map do |nodes|
```

File: app/serializers/serialize_exercise_instructions.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        task_title = parse_title(nodes.first)
+        task_text = nodes.drop(1).each.map(&:to_html).join.strip
+        task_hints = hints[task_title.downcase].to_a
+
+        tasks << { title: task_title, text: task_text, hints: task_hints }
```
Comment (iHiD) [inline app/serializers/serialize_exercise_instructions.rb]:
```suggestion
        { title: task_title, text: task_text, hints: task_hints }
```

File: app/serializers/serialize_exercise_instructions.rb [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+      tasks: tasks
+    }
+  end
+
+  private
+  def overview
+    instructions_doc.each.
+      take_while { |node| node.type != :header }.
+      map(&:to_html).
+      join
```
Comment (ErikSchierboom) [inline app/serializers/serialize_exercise_instructions.rb:30]: This has a small hidden bug, as just concatenating the HTML results in spacing being lost. I have to check how the HTML would be rendered when the elements are combined.

File: app/serializers/serialize_exercise_instructions.rb (hunk 5 above)
Comment (iHiD) [inline app/serializers/serialize_exercise_instructions.rb]: Better to not mutate the array. I can't remember if it's `1..-1`, or `1, -1`
```suggestion
        task_text = nodes[1,-1].each.map(&:to_html).join.strip
```

Comment (ErikSchierboom) [reply app/serializers/serialize_exercise_instructions.rb]: Oh, that mutates? I thought that a mutating method would have had a `!` postfix. I'll fix it

Comment (ErikSchierboom) [reply app/serializers/serialize_exercise_instructions.rb]: `[1..]` is the most concise version (suggested by Rubocop over `[1..-1]`.

Comment (iHiD) [reply app/serializers/serialize_exercise_instructions.rb]: That's nice. Didn't know of that.

File: app/commands/markdown/parse.rb [hunk 9]
```diff
@@ -32,18 +32,9 @@ def sanitized_html
...
-
-  memoize
-  def preprocessed_text
-    text.gsub(/^`{3,}(.*?)`{3,}\s*$/m) { "\n#{$&}\n" }
+    preprocessed_text = Markdown::Preprocess.(text)
+    doc = Markdown::RenderDoc.(preprocessed_text)
+    Markdown::Render.(doc, nofollow_links)
   end
 
   class Renderer < CommonMarker::HtmlRenderer
```
Comment (iHiD) [inline app/commands/markdown/parse.rb]: Can this be removed from here now?

Comment (ErikSchierboom) [reply app/commands/markdown/parse.rb]: Yes. Done

File: app/commands/markdown/preprocess.rb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+class Markdown::Preprocess
```
Comment (ErikSchierboom) [inline app/commands/markdown/preprocess.rb:1]: I've not added tests for this particular command as it is indirectly tested via the tests for the `Markdown::Parse` command. The same goes for the other extracted commands

File: app/javascript/components/editor/Header.tsx [hunk 11]
```diff
@@ -25,13 +26,11 @@ Header.Title = ({
   <div className="title">
     <div className="track">{trackTitle}</div>
     <div className="divider">/</div>
-    <div className="exeercise">{exerciseTitle}</div>
+    <div className="exercise">{exerciseTitle}</div>
```
Comment (ErikSchierboom) [inline app/javascript/components/editor/Header.tsx:29]: This seemed to be a typo

File: app/javascript/components/editor/Header.tsx [hunk 12]
```diff
@@ -25,13 +26,11 @@ Header.Title = ({
...
   </div>
 )
 
-Header.ActionHints = () => (
-  <button className="btn-small hints-btn">Hints</button>
-)
+Header.ActionHints = Hints
```
Comment (ErikSchierboom) [inline app/javascript/components/editor/Header.tsx:33]: As the hints display is now more involved, I've extracted it to a separate component

File: app/javascript/components/modals/BugReportModal.tsx [hunk 13]
```diff
@@ -75,7 +75,7 @@ export const BugReportModal = ({
   }, [open])
 
   return (
-    <Modal isOpen={open} onRequestClose={onClose} {...props}>
+    <Modal open={open} onClose={onClose} {...props}>
```
Comment (ErikSchierboom) [inline app/javascript/components/modals/BugReportModal.tsx]: This change is due to me having introduced our own `Modal` component that wraps the `react-modal` version.

File: app/javascript/components/modals/Modal.tsx [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+  children,
+  ...props
+}: {
+  open: boolean
+  onClose: () => void
+  children?: React.ReactNode
+}): JSX.Element => {
+  useLayoutEffect(() => {
+    // Required for accessibility: http://reactcommunity.org/react-modal/accessibility/#app-element
+    ReactModal.setAppElement(document.querySelector('body') as HTMLElement)
```
Comment (ErikSchierboom) [inline app/javascript/components/modals/Modal.tsx:18]: This is the main reason why I introduced this component, as we need to set the app element once the DOM is ready.

File: app/javascript/packs/application.tsx [hunk 15]
```diff
@@ -147,7 +147,9 @@ initReact({
       trackTitle={data.track_title}
       exerciseTitle={data.exercise_title}
       introduction={data.introduction}
-      instructions={data.instructions}
+      instructions={
+        (camelizeKeys(data.instructions) as unknown) as ExerciseInstructions
```
Comment (ErikSchierboom) [inline app/javascript/packs/application.tsx:152]: This was required to have the nested `generalHints` property be populated from the `general_hints` property

File: test/javascript/components/Editor/features/keyboard-shortcuts.test.js [hunk 16]
```diff
@@ -7,7 +7,10 @@ import { Editor } from '../../../../../app/javascript/components/Editor'
 
 test('toggle command palette when clicking on button', async () => {
   const { getByTitle, queryByText } = render(
-    <Editor files={[{ filename: 'file', content: 'file' }]} />
+    <Editor
+      files={[{ filename: 'file', content: 'file' }]}
+      instructions={{ overview: '', generalHints: [], tasks: [] }}
```
Comment (ErikSchierboom) [inline test/javascript/components/Editor/features/keyboard-shortcuts.test.js:12]: The editor now depends on these properties being provided

File: test/serializers/serialize_exercise_instructions_test.rb [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+class SerializeExerciseInstructionsTest < ActiveSupport::TestCase
+  test "serialize general hints" do
+    track = create :track, slug: 'fsharp'
+    exercise = create :concept_exercise, track: track, slug: 'log-levels'
+
+    serialized = SerializeExerciseInstructions.(exercise)
+
+    expected = [
+      "<p>The <code>string</code> class has many useful <a href=\"https://docs.microsoft.com/en-us/dotnet/api/system.string?view=netcore-3.1#methods\" target=\"_blank\">built-in\nmethods</a>.</p>\n", # rubocop:disable Layout/LineLength
+      "<p>Remember that strings are immutable.</p>\n"
```
Comment (ErikSchierboom) [inline test/serializers/serialize_exercise_instructions_test.rb:12]: We currently have a `\n` at the end, which is how the Markdown is rendered. I _could_ strip that off if you'd like that

File: app/commands/parse_markdown.rb (hunk 2 above)
Comment (SleeplessByte) [reply app/commands/parse_markdown.rb]: `Regexp.last_match(0)` is preferred over `$&`.

`$&` is a [special global variable](https://docs.ruby-lang.org/en/2.5.0/Regexp.html#class-Regexp-label-Special+global+variables) so you "must know" what it is.

File: app/serializers/serialize_exercise_instructions.rb (hunk 5 above)
Comment (SleeplessByte) [reply app/serializers/serialize_exercise_instructions.rb]: `.drop` does not mutate, but `[1..]` is something I see much more in the wild, so probably what you want here.

File: app/commands/markdown/preprocess.rb (hunk 10 above)
Comment (iHiD) [reply app/commands/markdown/preprocess.rb:1]: Cool. I might add these when I do the styling.

File: app/javascript/components/modals/BugReportModal.tsx (hunk 13 above)
Comment (iHiD) [reply app/javascript/components/modals/BugReportModal.tsx]: Nice. This'll make @kntsoriano happy!

File: test/serializers/serialize_exercise_instructions_test.rb (hunk 17 above)
Comment (iHiD) [reply test/serializers/serialize_exercise_instructions_test.rb:12]: I think it's fine as it is.

**Comment by iHiD** [comment]
I think this is great. I'll add styling tomorrow and then we can ship it before the call! Great work, mate :)

**Review by kntsoriano — APPROVED** [review]
Nice! LGTM

File: app/commands/parse_markdown.rb (hunk 2 above)
Comment (ErikSchierboom) [reply app/commands/parse_markdown.rb]: Ah, thanks

File: app/serializers/serialize_exercise_instructions.rb (hunk 5 above)
Comment (ErikSchierboom) [reply app/serializers/serialize_exercise_instructions.rb]: This has been fixed

File: app/javascript/components/modals/BugReportModal.tsx (hunk 13 above)
Comment (ErikSchierboom) [reply app/javascript/components/modals/BugReportModal.tsx]: I hope so!

**Comment by ErikSchierboom** [comment]
> I think this is great. I'll add styling tomorrow and then we can ship it before the call! Great work, mate :)

🥳

**Pull request merged**